File: src/dto/format.rs
use std::collections::HashMap;
use std::path::Path;

use super::items::{BaseItemDto, MediaStreamInfo, UserItemDataDto};

/// Format a database item row into an Emby-compatible BaseItemDto.
/// `item` is a sqlx::Row or a struct with the relevant fields.
pub fn format_item_dto(
    item: &ItemRow,
    server_id: &str,
    user_data: Option<&UserDataRow>,
) -> BaseItemDto {
    let mut dto = BaseItemDto {
        id: item.id.clone(),
        name: item.name.clone(),
        server_id: server_id.to_string(),
        item_type: item.item_type.clone(),
        sort_name: Some(item.sort_name.clone().unwrap_or_else(|| item.name.clone())),
        ..Default::default()
    };

    match item.item_type.as_str() {
        "Movie" | "Episode" => {
            dto.media_type = Some("Video".to_string());
            dto.is_folder = Some(false);
        }
        "Series" | "Season" | "CollectionFolder" => {
            dto.is_folder = Some(true);
        }
        _ => {}
    }

    dto.collection_type = item.collection_type.clone();
    dto.overview = item.overview.clone();
    dto.production_year = item.production_year;
    dto.premiere_date = item
        .premiere_date
        .as_ref()
        .map(|d| d.and_utc().to_rfc3339());
    dto.community_rating = item.community_rating;
    dto.official_rating = item.official_rating.clone();
    dto.run_time_ticks = item.runtime_ticks;
    dto.index_number = item.index_number;
    dto.parent_index_number = item.parent_index_number;
    dto.parent_id = item.parent_id.clone();
    dto.series_id = item.series_id.clone();
    dto.series_name = item.series_name.clone();
    dto.season_id = item.season_id.clone();
    dto.provider_ids = item.provider_ids.clone();

    // Path: prefer resolved_path, then try resolving .strm on-the-fly, then file_path
    let display_path = if let Some(ref rp) = item.resolved_path {
        Some(rp.clone())
    } else if let Some(ref fp) = item.file_path {
        if fp.ends_with(".strm") {
            resolve_strm_for_display(fp).or_else(|| Some(fp.clone()))
        } else {
            Some(fp.clone())
        }
    } else {
        None
    };
    dto.path = display_path.clone();

    // Container: use real container, not 'strm'
    if let Some(ref c) = item.container {
        if c != "strm" {
            dto.container = Some(c.clone());
        } else if let Some(ref p) = display_path {
            let ext = Path::new(p)
                .extension()
                .and_then(|e| e.to_str())
                .unwrap_or("mkv")
                .to_lowercase();
            dto.container = Some(ext);
        } else {
            dto.container = Some(c.clone());
        }
    }

    // Image tags
    let mut image_tags = HashMap::new();
    let series_item_id = item
        .series_fallback_id
        .clone()
        .or_else(|| item.series_id.clone())
        .or_else(|| item.parent_id.clone());

    if let Some(ref tag) = item.primary_image_tag {
        image_tags.insert("Primary".to_string(), tag.clone());
    }
    // For Episode/Season without own image: DON'T put series tag in ImageTags
    // Players use ImageTags.Primary + item's own ID to request images.
    // Instead, use SeriesPrimaryImageTag + SeriesId (set below in flat fields section).
    if !image_tags.is_empty() {
        dto.image_tags = Some(image_tags);
    }

    if let Some(ref tag) = item.backdrop_image_tag {
        dto.backdrop_image_tags = Some(vec![tag.clone()]);
    } else if item.item_type == "Episode" || item.item_type == "Season" {
        if let Some(ref tag) = item.series_backdrop_image_tag {
            // Don't set backdrop_image_tags on the episode itself
            // Use ParentBackdrop fields instead
            dto.parent_backdrop_item_id = series_item_id.clone();
            dto.parent_backdrop_image_tags = Some(vec![tag.clone()]);
        }
    }

    // Emby player compatibility: flat image fields for Episode/Season
    if item.item_type == "Episode" || item.item_type == "Season" {
        if let Some(ref tag) = item.series_primary_image_tag {
            dto.series_primary_image_tag = Some(tag.clone());
            if dto.series_primary_image_item_id.is_none() {
                dto.series_primary_image_item_id = series_item_id.clone();
            }
            // ParentPrimaryImage for season/episode navigation
            dto.parent_primary_image_item_id = series_item_id.clone();
            dto.parent_primary_image_tag = Some(tag.clone());
            // ParentThumb (players use this for episode thumbnails)
            dto.parent_thumb_item_id = series_item_id.clone();
            dto.parent_thumb_image_tag = Some(tag.clone());
        }
        if let Some(ref tag) = item.series_backdrop_image_tag {
            if dto.parent_backdrop_item_id.is_none() {
                dto.parent_backdrop_item_id = series_item_id.clone();
                dto.parent_backdrop_image_tags = Some(vec![tag.clone()]);
            }
        }
    }

    dto.child_count = item.child_count;
    dto.recursive_item_count = item.recursive_item_count;

    // User data
    dto.user_data = Some(if let Some(ud) = user_data {
        let position = ud.playback_position_ticks.unwrap_or(0);
        let play_count = ud.play_count.unwrap_or(0);
        let is_fav = ud.is_favorite.unwrap_or(false);
        let played = ud.played.unwrap_or(false);
        let percentage = if let Some(runtime) = dto.run_time_ticks {
            if runtime > 0 && position > 0 {
                Some((position as f64 / runtime as f64) * 100.0)
            } else {
                None
            }
        } else {
            None
        };
        UserItemDataDto {
            playback_position_ticks: position,
            play_count,
            is_favorite: is_fav,
            played,
            last_played_date: ud
                .last_played_date
                .as_ref()
                .map(|d| d.and_utc().to_rfc3339()),
            played_percentage: percentage,
        }
    } else {
        UserItemDataDto {
            playback_position_ticks: 0,
            play_count: 0,
            is_favorite: false,
            played: false,
            last_played_date: None,
            played_percentage: None,
        }
    });

    dto
}

/// Resolve a .strm file to its actual media path for display in DTOs.
/// Reads the first line of the strm file. For remote URLs returns as-is.
/// For local paths, applies mount path fixups.
fn resolve_strm_for_display(strm_path: &str) -> Option<String> {
    let content = std::fs::read_to_string(strm_path).ok()?;
    let line = content.lines().next()?.trim();
    if line.is_empty() || line.starts_with('#') {
        return None;
    }
    let mut resolved = line.to_string();
    if resolved.starts_with("http") {
        return Some(resolved);
    }
    if !resolved.starts_with('/') {
        return None;
    }
    if !Path::new(&resolved).exists() {
        let mnt_path = format!("/mnt{resolved}");
        if Path::new(&mnt_path).exists() {
            resolved = mnt_path;
        } else {
            let fixed = resolved.replacen("/CloudNAS", "/mnt/CloudNAS", 1);
            if fixed != resolved && Path::new(&fixed).exists() {
                resolved = fixed;
            }
        }
    }
    Some(resolved)
}

pub fn format_media_stream_dto(stream: &StreamRow) -> MediaStreamInfo {
    MediaStreamInfo {
        codec: stream.codec.clone().unwrap_or_default(),
        stream_type: stream.stream_type.clone(),
        index: stream.stream_index,
        language: stream.language.clone(),
        title: stream.title.clone(),
        is_default: stream.is_default.unwrap_or(false),
        is_forced: stream.is_forced.unwrap_or(false),
        width: stream.width,
        height: stream.height,
        bit_rate: stream.bit_rate,
        channels: stream.channels,
        sample_rate: stream.sample_rate,
        bit_depth: stream.bit_depth,
        pixel_format: stream.pixel_format.clone(),
        display_title: stream.display_title.clone(),
    }
}

// Row types that map to database columns
// These are used by format functions and will be populated by sqlx queries

#[derive(Debug, Clone, Default)]
pub struct ItemRow {
    pub id: String,
    pub name: String,
    pub item_type: String,
    pub sort_name: Option<String>,
    pub collection_type: Option<String>,
    pub overview: Option<String>,
    pub production_year: Option<i32>,
    pub premiere_date: Option<chrono::NaiveDateTime>,
    pub community_rating: Option<f64>,
    pub official_rating: Option<String>,
    pub runtime_ticks: Option<i64>,
    pub index_number: Option<i32>,
    pub parent_index_number: Option<i32>,
    pub parent_id: Option<String>,
    pub series_id: Option<String>,
    pub series_name: Option<String>,
    pub season_id: Option<String>,
    pub container: Option<String>,
    pub file_path: Option<String>,
    pub resolved_path: Option<String>,
    pub provider_ids: Option<serde_json::Value>,
    pub primary_image_tag: Option<String>,
    pub backdrop_image_tag: Option<String>,
    pub series_primary_image_tag: Option<String>,
    pub series_backdrop_image_tag: Option<String>,
    pub series_fallback_id: Option<String>,
    pub child_count: Option<i64>,
    pub recursive_item_count: Option<i64>,
}

#[derive(Debug, Clone, Default)]
pub struct UserDataRow {
    pub playback_position_ticks: Option<i64>,
    pub play_count: Option<i32>,
    pub is_favorite: Option<bool>,
    pub played: Option<bool>,
    pub last_played_date: Option<chrono::NaiveDateTime>,
}

#[derive(Debug, Clone, Default)]
pub struct StreamRow {
    pub codec: Option<String>,
    pub stream_type: String,
    pub stream_index: i32,
    pub language: Option<String>,
    pub title: Option<String>,
    pub is_default: Option<bool>,
    pub is_forced: Option<bool>,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub bit_rate: Option<i64>,
    pub channels: Option<i32>,
    pub sample_rate: Option<i32>,
    pub bit_depth: Option<i32>,
    pub pixel_format: Option<String>,
    pub display_title: Option<String>,
}

File: internal/dto/format.go
package dto

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync/atomic"

	"github.com/google/uuid"
)

// strmItemPathResolvedFlag 控制 strm 条目「item 级 Path」字段的取值:
//   - false(默认)= 返回 .strm 文件本身的路径(对齐 Emby:item.Path 永远是 .strm,
//     解析后的真实地址只出现在 MediaSources.Path)
//   - true        = 返回解析 .strm 后的内层真实路径(FYMS 旧行为)
//
// 无论该开关取值如何,strm 的 item.Container 都解析为真实内层容器(如 mp4),与 Emby 一致。
// 由 main.go 启动时按 system_config 'strm_item_path_mode' 设置,postConfiguration 保存时实时刷新。
var strmItemPathResolvedFlag atomic.Bool

// SetStrmItemPathMode 设置 strm item.Path 模式。mode=="resolved" 时返回解析后真实路径,
// 其它值(含默认空值 / "strm")返回 .strm 文件路径。
func SetStrmItemPathMode(mode string) {
	strmItemPathResolvedFlag.Store(strings.EqualFold(strings.TrimSpace(mode), "resolved"))
}

func strmItemPathResolved() bool {
	return strmItemPathResolvedFlag.Load()
}

// StrmItemPathResolved 导出当前模式:true=item.Path 取解析后真实路径;false=取 .strm 文件路径(默认)。
// 供 handler 层判断是否应把 item 级 Path 覆盖为 MediaSources 的解析路径。
func StrmItemPathResolved() bool {
	return strmItemPathResolvedFlag.Load()
}

// studioNamespace 用于由 studio name 生成稳定 UUID。Emby/Jellyfin 客户端
// （包括 VidHub）要求 Studios[].Id 必填且为 UUID 形式。
var studioNamespace = uuid.MustParse("b2c3d4e5-f6a7-4b8c-9d0e-1f2a3b4c5d6e")

func studioStableID(name string) string {
	return uuid.NewSHA1(studioNamespace, []byte(name)).String()
}

// normalizeProviderIDs 把 DB 里统一的小写 provider key（tmdb/imdb/tvdb...）在出口处
// 补出 Emby 官方大小写 key（Tmdb/Imdb/Tvdb...），同时保留小写 key。
// 这样取 "Imdb" 的聚合 app 和取 "imdb" 的 app 都能拿到值，多余 key 主流客户端忽略。
// DB 存储层保持小写不变，只在序列化时转换。任何解析/序列化失败都原样返回，绝不丢数据。
func normalizeProviderIDs(raw *json.RawMessage) *json.RawMessage {
	if raw == nil || len(*raw) == 0 {
		return raw
	}
	var m map[string]string
	if err := json.Unmarshal(*raw, &m); err != nil || len(m) == 0 {
		return raw
	}
	out := make(map[string]string, len(m)*2)
	for k, v := range m {
		lk := strings.ToLower(strings.TrimSpace(k))
		if lk == "" || v == "" {
			continue
		}
		out[lk] = v                    // 保留小写
		out[canonicalProvider(lk)] = v // 补官方大小写
	}
	if len(out) == 0 {
		return raw
	}
	b, err := json.Marshal(out)
	if err != nil {
		return raw
	}
	rm := json.RawMessage(b)
	return &rm
}

// canonicalProvider 已知 provider 用 Emby 官方写法，未知的首字母大写兜底。
func canonicalProvider(lower string) string {
	switch lower {
	case "tmdb":
		return "Tmdb"
	case "imdb":
		return "Imdb"
	case "tvdb":
		return "Tvdb"
	case "tmdbcollection":
		return "TmdbCollection"
	default: // bangumi→Bangumi, douban→Douban 等
		if lower == "" {
			return lower
		}
		return strings.ToUpper(lower[:1]) + lower[1:]
	}
}

// FormatItemDtoList 列表场景：跳过 strm 文件解析（避免大量磁盘 IO）
func FormatItemDtoList(item *ItemRow, serverID string, userData *UserDataRow) BaseItemDto {
	return formatItemDto(item, serverID, userData, true)
}

func FormatItemDto(item *ItemRow, serverID string, userData *UserDataRow) BaseItemDto {
	return formatItemDto(item, serverID, userData, false)
}

func formatItemDto(item *ItemRow, serverID string, userData *UserDataRow, skipStrmResolve bool) BaseItemDto {
	sortName := item.Name
	if item.SortName != nil {
		sortName = *item.SortName
	}

	dto := BaseItemDto{
		ID:       item.ID,
		Name:     item.Name,
		ServerID: serverID,
		Type:     item.ItemType,
		SortName: &sortName,
	}

	switch item.ItemType {
	case "Movie", "Episode":
		mediaType := "Video"
		isFolder := false
		dto.MediaType = &mediaType
		dto.IsFolder = &isFolder
	case "Series", "Season", "CollectionFolder", "Folder":
		isFolder := true
		dto.IsFolder = &isFolder
	}

	dto.CollectionType = item.CollectionType
	dto.Overview = item.Overview
	dto.ProductionYear = item.ProductionYear
	if item.PremiereDate != nil {
		s := item.PremiereDate.UTC().Format("2006-01-02T15:04:05.0000000Z")
		dto.PremiereDate = &s
	}
	dto.CommunityRating = item.CommunityRating
	dto.OfficialRating = item.OfficialRating
	dto.RunTimeTicks = item.RuntimeTicks
	dto.IndexNumber = item.IndexNumber
	dto.ParentIndexNumber = item.ParentIndexNumber
	dto.ParentID = item.ParentID
	dto.SeriesID = item.SeriesID
	dto.SeriesName = item.SeriesName
	dto.SeasonID = item.SeasonID
	dto.ProviderIDs = normalizeProviderIDs(item.ProviderIDs)

	// displayPath = item 级 Path 字段；resolvedStrm = 解析 .strm 得到的内层真实路径
	// (仅非列表模式下读盘获得，用于推导真实 Container)。
	var displayPath *string
	var resolvedStrm *string
	if item.ResolvedPath != nil {
		displayPath = item.ResolvedPath
	} else if item.FilePath != nil {
		isStrm := strings.HasSuffix(strings.ToLower(*item.FilePath), ".strm")
		switch {
		case skipStrmResolve:
			// 列表模式：跳过磁盘 IO，item.Path 直接用 file_path(.strm 时即对齐 Emby)。
			displayPath = item.FilePath
		case isStrm:
			// 读一次 .strm 取内层真实路径，供 Container 推导；
			// item.Path 用 .strm 还是解析后路径由 strmItemPathResolved() 决定。
			resolvedStrm = resolveStrmForDisplay(*item.FilePath)
			if strmItemPathResolved() && resolvedStrm != nil {
				displayPath = resolvedStrm
			} else {
				displayPath = item.FilePath
			}
		default:
			displayPath = item.FilePath
		}
	}
	dto.Path = displayPath

	if item.Container != nil {
		if *item.Container != "strm" {
			dto.Container = item.Container
		} else {
			// items.container 对 strm 入库时存的是字面 "strm"，这里推真实内层容器(对齐 Emby 的 mp4)。
			// 优先用解析后的内层路径扩展名；列表模式没解析时退回 displayPath，避免对外暴露 "strm" 容器。
			src := resolvedStrm
			if src == nil {
				src = displayPath
			}
			if src != nil {
				ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(*src), "."))
				if ext == "" || ext == "strm" {
					ext = "mkv"
				}
				dto.Container = &ext
			} else {
				dto.Container = item.Container
			}
		}
	}

	seriesItemID := item.SeriesFallbackID
	if seriesItemID == nil {
		seriesItemID = item.SeriesID
	}
	if seriesItemID == nil {
		seriesItemID = item.ParentID
	}

	imageTags := make(map[string]string)
	if item.PrimaryImageTag != nil {
		imageTags["Primary"] = *item.PrimaryImageTag
	}
	if len(imageTags) > 0 {
		dto.ImageTags = imageTags
	}

	if item.BackdropImageTag != nil {
		dto.BackdropImageTags = []string{*item.BackdropImageTag}
	} else if item.ItemType == "Episode" || item.ItemType == "Season" {
		if item.SeriesBackdropImageTag != nil {
			dto.ParentBackdropItemID = seriesItemID
			dto.ParentBackdropImageTags = []string{*item.SeriesBackdropImageTag}
		}
	}

	if item.ItemType == "Episode" || item.ItemType == "Season" {
		if item.SeriesPrimaryImageTag != nil {
			dto.SeriesPrimaryImageTag = item.SeriesPrimaryImageTag
			if dto.SeriesPrimaryImageItemID == nil {
				dto.SeriesPrimaryImageItemID = seriesItemID
			}
			dto.ParentPrimaryImageItemID = seriesItemID
			dto.ParentPrimaryImageTag = item.SeriesPrimaryImageTag
			dto.ParentThumbItemID = seriesItemID
			dto.ParentThumbImageTag = item.SeriesPrimaryImageTag
		}
		if item.SeriesBackdropImageTag != nil {
			if dto.ParentBackdropItemID == nil {
				dto.ParentBackdropItemID = seriesItemID
				dto.ParentBackdropImageTags = []string{*item.SeriesBackdropImageTag}
			}
		}
	}

	dto.ChildCount = item.ChildCount
	dto.RecursiveItemCount = item.RecursiveItemCount

	// Supplemental fields for bot/search compatibility
	if item.Tagline != nil && *item.Tagline != "" {
		dto.Taglines = []string{*item.Tagline}
	}
	if item.CreatedAt != nil {
		t := item.CreatedAt.UTC().Format("2006-01-02T15:04:05.0000000Z")
		dto.DateCreated = &t
	}
	if item.Studio != nil && *item.Studio != "" {
		dto.Studios = []StudioItem{{Name: *item.Studio, ID: studioStableID(*item.Studio)}}
	}
	dto.ProductionLocations = []string{}

	if userData != nil {
		position := int64(0)
		if userData.PlaybackPositionTicks != nil {
			position = *userData.PlaybackPositionTicks
		}
		playCount := int32(0)
		if userData.PlayCount != nil {
			playCount = *userData.PlayCount
		}
		isFav := false
		if userData.IsFavorite != nil {
			isFav = *userData.IsFavorite
		}
		played := false
		if userData.Played != nil {
			played = *userData.Played
		}

		var percentage *float64
		if dto.RunTimeTicks != nil && *dto.RunTimeTicks > 0 && position > 0 {
			p := float64(position) / float64(*dto.RunTimeTicks) * 100.0
			percentage = &p
		}

		var lastPlayed *string
		if userData.LastPlayedDate != nil {
			s := userData.LastPlayedDate.UTC().Format("2006-01-02T15:04:05.0000000Z")
			lastPlayed = &s
		}

		dto.UserData = &UserItemDataDto{
			PlaybackPositionTicks: position,
			PlayCount:             playCount,
			IsFavorite:            isFav,
			Played:                played,
			LastPlayedDate:        lastPlayed,
			PlayedPercentage:      percentage,
		}
	} else {
		dto.UserData = &UserItemDataDto{}
	}

	return dto
}

func FormatMediaStreamDto(stream *StreamRow) MediaStreamInfo {
	codec := ""
	if stream.Codec != nil {
		codec = *stream.Codec
	}
	return MediaStreamInfo{
		Codec:        codec,
		Type:         stream.StreamType,
		Index:        stream.StreamIndex,
		Language:     stream.Language,
		Title:        stream.Title,
		IsDefault:    stream.IsDefault != nil && *stream.IsDefault,
		IsForced:     stream.IsForced != nil && *stream.IsForced,
		Width:        stream.Width,
		Height:       stream.Height,
		BitRate:      stream.BitRate,
		Channels:     stream.Channels,
		SampleRate:   stream.SampleRate,
		BitDepth:     stream.BitDepth,
		PixelFormat:  stream.PixelFormat,
		DisplayTitle: stream.DisplayTitle,
	}
}

func resolveStrmForDisplay(strmPath string) *string {
	data, err := os.ReadFile(strmPath)
	if err != nil {
		return nil
	}
	lines := strings.SplitN(string(data), "\n", 2)
	if len(lines) == 0 {
		return nil
	}
	line := strings.TrimSpace(lines[0])
	if line == "" || strings.HasPrefix(line, "#") {
		return nil
	}

	resolved := line
	if strings.HasPrefix(resolved, "http") {
		return &resolved
	}
	if !strings.HasPrefix(resolved, "/") {
		return nil
	}

	if _, err := os.Stat(resolved); err != nil {
		mntPath := "/mnt" + resolved
		if _, err := os.Stat(mntPath); err == nil {
			return &mntPath
		}
		fixed := strings.Replace(resolved, "/CloudNAS", "/mnt/CloudNAS", 1)
		if fixed != resolved {
			if _, err := os.Stat(fixed); err == nil {
				return &fixed
			}
		}
	}
	return &resolved
}
